src/Raven.Server/Documents/Indexes/IndexStore.cs

```diff
@@ -57,40 +56,103 @@ public class IndexStore : IDisposable
             _serverStore = serverStore;
-            _logger = LoggingSource.Instance.GetLogger<IndexStore>(_documentDatabase.Name);
+            Logger = LoggingSource.Instance.GetLogger<IndexStore>(_documentDatabase.Name);
 
             var stoppedConcurrentIndexBatches = _documentDatabase.Configuration.Indexing.NumberOfConcurrentStoppedBatchesIfRunningLowOnMemory;
             StoppedConcurrentIndexBatches = new SemaphoreSlim(stoppedConcurrentIndexBatches);
         }
 
-        public void HandleDatabaseRecordChange(DatabaseRecord record, long index)
+        public void HandleDatabaseRecordChange(DatabaseRecord record, long raftIndex)
         {
             if (record == null)
                 return;
 
-            HandleSorters(record, index);
-            HandleDeletes(record, index);
-            HandleChangesForStaticIndexes(record, index);
-            HandleChangesForAutoIndexes(record, index);
+            var indexesToStart = new List<Index>();
+
+            HandleSorters(record, raftIndex);
+            HandleDeletes(record, raftIndex);
+
+            HandleChangesForStaticIndexes(record, raftIndex, indexesToStart);
+            HandleChangesForAutoIndexes(record, raftIndex, indexesToStart);
+
+            if (indexesToStart.Count <= 0)
+                return;
+
+            var sp = Stopwatch.StartNew();
+
+            if (Logger.IsInfoEnabled)
+                Logger.Info($"Starting {indexesToStart.Count} indexes");
+
+            ExecuteForIndexes(indexesToStart, index =>
+            {
+                var indexLock = GetIndexLock(index.Name);
+
+                try
+                {
+                    indexLock.Wait(_documentDatabase.DatabaseShutdown);
+                }
+                catch (OperationCanceledException e)
+                {
+                    _documentDatabase.RachisLogIndexNotifications.NotifyListenersAbout(raftIndex, e);
+                    return;
+                }
+
+                try
+                {
+                    StartIndex(index);
+                }
+                catch (Exception e)
+                {
+                    _documentDatabase.RachisLogIndexNotifications.NotifyListenersAbout(raftIndex, e);
+                    if (Logger.IsInfoEnabled)
+                        Logger.Info($"Could not start index `{index.Name}`", e);
+                }
+                finally
+                {
+                    indexLock.Release();
+                }
+            });
+
+            if (Logger.IsInfoEnabled)
+                Logger.Info($"Started {indexesToStart.Count} indexes, took: {sp.ElapsedMilliseconds}ms");
+        }
+
+        private void ExecuteForIndexes(IEnumerable<Index> indexes, Action<Index> action)
+        {
+            int numberOfUtilizedCores;
+            using (_documentDatabase.ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
```

Comment:
I do not think that this context is needed?

Reply:
removed